examples/randlanet_classification.py

```diff
@@ -0,0 +1,256 @@
+
+        Args:
+            x_j (Tensor): neighboors features (K,d)
+            pos_i (Tensor): centroid position (repeated) (K,3)
+            pos_j (Tensor): neighboors positions (K,3)
+
+        returns:
+            (Tensor): locSE weighted by feature attention scores.
+
+        """
+        # Encode local neighboorhod structural information
+        pos_diff = pos_j - pos_i
+        distance = torch.sqrt((pos_diff * pos_diff).sum(1, keepdim=True))
+        relative_infos = torch.cat(
+            [pos_i, pos_j, pos_diff, distance], dim=1
+        )  # N//decim * K, d
+        local_spatial_encoding = self.mlp_encoder(relative_infos)  # N//decim * K, d
+        local_features = torch.cat(
+            [x_j, local_spatial_encoding], dim=1
+        )  # N//decim * K, 2d
+
+        # Attention will weight the different features of x
+        att_features = self.mlp_attention(local_features)  # N//decim * K, d_out
+        d_out = att_features.size(1)
+        att_scores = torch.softmax(
+            att_features.view(-1, self.num_neighbors, d_out), dim=1
+        ).view(
+            -1, d_out
+        )  # N//decim * K, d_out with transitory N//decim, K, d_out
+
+        return att_scores * local_features  # N//decim * K, d_out
+
+
+class DilatedResidualBlock(MessagePassing):
+    def __init__(
+        self,
+        decimation,
+        num_neighbors,
+        d_in: int,
+        d_out: int,
+    ):
+        super().__init__()
+        self.num_neighbors = num_neighbors
+        self.decimation = decimation
+        self.d_in = d_in
+        self.d_out = d_out
+
+        # MLP on input
+        self.mlp1 = default_MLP([d_in, d_out // 8])
+        # MLP on input, and the result is summed with the output of mlp2
+        self.shortcut = default_MLP([d_in, d_out], act=None)
+        # MLP on output
+        self.mlp2 = default_MLP([d_out // 2, d_out], act=None)
+
+        self.lfa1 = LocalFeatureAggregation(d_out // 4, num_neighbors)
+        self.lfa2 = LocalFeatureAggregation(d_out // 2, num_neighbors)
+
+        self.lrelu = torch.nn.LeakyReLU(**lrelu02_kwargs)
+
+    def forward(self, x, pos, batch, return_unsampled_as_well: bool = False):
+        idx = subsample_by_decimation(batch, self.decimation)
+        row, col = knn(pos, pos, self.num_neighbors, batch_x=batch, batch_y=batch)
+        edge_index = torch.stack([col, row], dim=0)
+
+        shortcut_of_x = self.shortcut(x)  # N, d_out
+        x = self.mlp1(x)  # N, d_out//8
+        x = self.lfa1(edge_index, x, pos)  # N, d_out//2
+        x = self.lfa2(edge_index, x, pos)  # N, d_out//2
+        x = self.mlp2(x)  # N, d_out
+        x = self.lrelu(x + shortcut_of_x)  # N, d_out
+
+        sampled = (x[idx], pos[idx], batch[idx])
+        if return_unsampled_as_well:
+            # Needed for skip connection in final upsampling
+            return sampled, (x, pos, batch)
+        return sampled  # N//decim, d_out
+
+
+def subsample_by_decimation(batch, decimation):
```

Comment:
Good idea as well. This makes the main network forward somewhat more complex, but with the advantage to avoid an edge case where we return the unsampled block1 output, and also a simpler decimation function.
I cherry-picked some lines in your implementation, thank you very much for sharing it :)